### ISSUE TITLE
Review HttpClient backpressure semantic

### ISSUE THREAD

**Issue by sbordet** [eclipse/jetty.project#2429]
Currently, when `HttpClient` receives response content backpressure is applied by means of succeeding the callback passed to `AsyncContentListener.onContent(Response, ByteBuffer, Callback)`.

Succeeding this callback has the meaning of telling to the implementation that the `ByteBuffer` has been consumed (and therefore can be recycled) - the _release_ semantic, **and** the meaning of demanding another `ByteBuffer` from the implementation - the _demand_ semantic.

There are cases where the 2 meanings should be separate. 
For example, demanding more than 1 buffer to process them all together, and later releasing them all.
Or demanding all the buffers first, and only later process them and releasing them.

See also: https://jira.spring.io/browse/SPR-15092 and https://github.com/spring-projects/spring-framework/pull/1783.

**Comment by sbordet** [comment]
Separating the 2 semantics may be achieved without changing the user APIs, something like:

```java
onData(DataFrame frame, Callback callback) {
  BackPressureCallback bPCB = (BackPressureCallback)callback;
  if (!frame.isEndStream()) {
    frames.add(frame);
    callbacks.add(bPCB);
    bPCB.demand(1);
  } else {
    int totalLength = frames.stream()
            .mapToInt(f -> f.getData().remaining())
            .sum();
    callbacks.forEach(bPCB -> bPCB.release());
  }
}
```

Particular attention should be paid to notifying multiple application listeners: one listener may not call `demand(...)` while the other calls `demand(2)` - we need to track demands for each application listener separately.
We want to lockstep the notifications, i.e. all listeners should have at least called `demand(1)` before delivering the next chunk of content. One listener may delay all the others but I guess there is no way around it.
In ReactiveStreams world, [this blog](http://akarnokd.blogspot.it/2017/12/java-9-flow-api-multicasting-via.html) explains a `MulticastProcessor`, whose functionality (maybe without using the ReactiveStreams API to not make Jetty depend on ReactiveStreams) is exactly what is needed here.

**Comment by sbordet** [comment]
This issue and #2439 are closely related.

**Comment by gregw** [comment]
I don't think `demand` is the right semantic here.     It is essentially the reactive stream semantic, except that there is an implicit `demand(1)`.   I think it would be confusing to mimic the reactive style, but not have the same semantic.   I also don't think it is worthwhile passing a `long`, as it will always be `1`, plus the meaning of `>1` is pretty unknown as buffers can be of variable length.

Finally, while I know we've done it elsewhere, I really don't like typed callbacks.   Callbacks are frequently wrapped and/or aggregated, so typed call backs can quickly become a multi-inheritance nightmare.

I think the same separation of concerns can be achieved simply by adding `retain()` and `release()` methods/semantics to the DataFrame.  Calling the normal callback methods indicates the processing is complete and more calls to `onData` can be made.   Calling `frame.retain()` during the `onData(...)` call indicates that the data has been sliced/referenced and the buffer is not available to be reused until a matching `release()` call.

**Comment by joakime** [comment]
Merged in PR #4106
